framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -705,6 +698,35 @@ public Void visitMethod(MethodTree node, Void p) {
         }
     }
 
+    /**
+     * Issue a warning if the result type of the constructor is a subtype of the declaring class. If
+     * it is a supertype of the class, then a type.invalid.conflicting.annos error will be issued by
+     * {@link #isValidUse(AnnotatedDeclaredType, AnnotatedDeclaredType, Tree)}.
+     *
+     * @param constructorType AnnotatedExecutableType for the constructor
+     * @param constructorElement element that declares the constructor
+     */
+    protected void checkConstructorResult(
+            AnnotatedExecutableType constructorType, ExecutableElement constructorElement) {
+        Set<AnnotationMirror> constructorAnnotations =
+                constructorType.getReturnType().getAnnotations();
+        Set<AnnotationMirror> classAnnotations =
+                atypeFactory
+                        .getAnnotatedType(constructorElement.getEnclosingElement())
+                        .getAnnotations();
+        QualifierHierarchy qualifierHierarchy = atypeFactory.getQualifierHierarchy();
+        for (AnnotationMirror classAnno : classAnnotations) {
+            AnnotationMirror constructorAnno =
+                    qualifierHierarchy.findAnnotationInSameHierarchy(
+                            constructorAnnotations, classAnno);
+            if (!qualifierHierarchy.isSubtype(classAnno, constructorAnno)) {
```

Comment:
The javadoc says `Issue a warning if the result type of the constructor is a subtype of the declaring class.`.
Wouldn't that be more directly implemented as `if (qH.isSubtype(constuctorAnno, classAnno)`?
Ah, but we want to allow that they are the same. So maybe adjust the javadoc?

In the end, don't these rules enforce that the annotation on the constructor result is exactly the same as the one on the class declaration?
I thought the rule would be that a specific constructor could use a more specific type. Otherwise, why do we need annotations on constructor results at all?

Reply:
>So maybe adjust the javadoc?

Yes, I'll do that. 

>why do we need annotations on constructor results at all?

We are going to add something to the manual about this.  I should probably add something here or a pointer to the manual.  The manual section will say something like this:

Any annotation on a constructor result, other than that on the declaration of the class, will result in a warning.  For an arbitrary checker, this is the desired behavior because the Checker Framework can't verify that the constructor result annotation is correct.  Specific checkers may be able to verify these annotations and thus not issue the warning.

For example, constructor results are always non-null, so it is legal to annotation the constructor result as `@NonNull` even if the annotation on the class declaration is `@Nullable`.  In the Aliasing Checker, a constructor result can be annotated with `@Unique` because the checker issues a warning if the constructed object is leaked.

Reply:
I agree with this somewhat; however, we also have no way to verify that the constructor produces an object of the type on the class declaration, either. So in some situations this should also warn if there is no annotation on the constructor.

Let's take a qualifier hierarchy `@A <: @B <: @C:`
For a class `@B class Foo { Foo() {} }` there is no way for us to veryify that the constructor establishes that the object actually fulfills the property `@B`. Maybe the default only establishes `@C` and the `@B` on the class is too strong.
Nothing special gets enforced about the annotation on the class declaration, so I don't see how we can guarantee that the constructor establishes that property.
Does that make sense?

Reply:
You're saying that if a checker issues an `inconsistent.constructor.type` warning for `@A Foo(){}`, then it should also issue a warning for`@B Foo(){}`, right?  That makes sense to me.  Let's discuss with Mike tomorrow.

Note, a checker will issue a `super.invocation.invalid` warning for both constructors--unless Object was annotated with `@B`.

Reply:
Can you write the example `@B class Foo { Foo() {} }` without defaults?  Defaults don't affect the answer, so leaving them off is confusing.

I'm assuming the constructor result type is `@B`.  In that case,  then there is no need for a warning, because every `Foo` (including the newly-constructed one) is a `@B` due to the class annotation.

PS: A small request: please use the hierarchy `@C <: @B <: @A`.  It is a long-standing tradition to assign letters from the top down, so it is confusing to reverse it.

Reply:
Ok, switching to `@C <: @B <: @A`.
If I write `@A class Foo { @B Foo() {} }` I get a warning about the `inconsistent.constructor.type`.
If I write `@B class Foo { @B Foo() {} }` I do not get that warning.
However, the `@B` on the class declaration does not enforce anything about the constructed objects.
So why would the framework issue a warning for the first code and not the second? For neither code can the framework verify that the object is actually `@B`.

Reply:
We discussed this some more: if a constructor result annotation is anything other than the top qualifier in the hierarchy, we issue the `inconsistent.constructor.type` error.